lua/terrortown/menus/gamemode/base_gamemodemenu.lua

```diff
@@ -102,6 +106,60 @@ function CLGAMEMODEMENU:AddSubmenu(submenu)
 	self.submenus[#self.submenus + 1] = submenu
 end
 
+---
+-- Checks if the menu has a searchbar enabled.
+-- @note This function should be overwritten and return true, if you want a searchbar.
+-- @return boolean Return true if searchbar should be available
+-- @realm client
+function CLGAMEMODEMENU:HasSearchbar()
+	return false
+end
+
+---
+-- Filters the list with a searchText and returns full list if nothing is entered.
+-- @note This function can be overwritten for a custom method to use the searchBar.
+-- @param string searchText
+-- @return menuClasses Returns a list of all matching submenus, needs to be indexed with ascending numbers
+-- @realm client
+function CLGAMEMODEMENU:GetMatchingSubmenus(searchText)
+	local submenuClasses = self:GetVisibleSubmenus()
+
+	if searchText == "" then
+		return submenuClasses
+	end
+
+	local filteredSubmenuClasses = {}
+
+	local counter = 0
+	for i = 1, #submenuClasses do
+		local submenuClass = submenuClasses[i]
+
+		if self:MatchesSearchString(submenuClass, searchText) then
+			counter = counter + 1
+			filteredSubmenuClasses[counter] = submenuClass
+		end
+	end
+
+	return filteredSubmenuClasses
+end
+
+---
+-- Determines the used searchfunction.
+-- Parameters for that function are submenuClasses and the searchText
+-- Per default only titles are searched and compared to the searchtext in lowercase letters.
+-- @note This function can be overwritten to use a custom searchfunction.
+-- @param menuClass submenuClass
+-- @param string searchText
+-- @return bool Returns 
+-- @realm client
+function CLGAMEMODEMENU:MatchesSearchString(submenuClass, searchText)
+	local txt = stringLower(searchText)
+	local title = stringLower(TryT(submenuClass.title))
+	local start = stringFind(title, txt)
+
+	return start or false and true
+end
+
```

Comment:
Refactored that, I liked your idea and already found a major performance boost, if you dont search, you dont need to go through the complete for loop to get the same list as before 😄

Reply:
Personally I'd add a note to `CLGAMEMODEMENU:GetMatchingSubmenus` to make it clear to devs that they probably do not want to overwrite that function. Otherwise I really like it, great work!